patroni/config_generator.py

```diff
@@ -0,0 +1,466 @@
+
+        :raises:
+            :class:`PatroniException`: if :exc:`OSError` occured during the conf files handling.
+        """
+        default_hba_path = os.path.join(self.config['postgresql']['data_dir'], 'pg_hba.conf')
+        if self.config['postgresql']['parameters']['hba_file'] == default_hba_path:
+            try:
+                self.config['postgresql']['pg_hba'] = list(
+                    filter(lambda i: i and i.split()[0] in self._get_hba_conn_types, read_stripped(default_hba_path)))
+            except OSError as err:
+                raise PatroniException(f'Failed to read pg_hba.conf: {err}')
+
+        default_ident_path = os.path.join(self.config['postgresql']['data_dir'], 'pg_ident.conf')
+        if self.config['postgresql']['parameters']['ident_file'] == default_ident_path:
+            try:
+                self.config['postgresql']['pg_ident'] = [i for i in read_stripped(default_ident_path)
+                                                         if i and not i.startswith('#')]
+            except OSError as err:
+                raise PatroniException(f'Failed to read pg_ident.conf: {err}')
+            if not self.config['postgresql']['pg_ident']:
+                del self.config['postgresql']['pg_ident']
+
+    def _enrich_config_from_running_instance(self) -> None:
+        """Extend `self.config` dictionary with the values gathered from a running instance.
+
+        Retrieve the following information from the running PostgreSQL instance:
+
+        * superuser auth parameters (see :func:`~RunningClusterConfigGenerator._set_su_params`)
+        * some GUC values (see :func:`~RunningClusterConfigGenerator._set_pg_params`)
+        * ``postgresql.connect_address``, ``postgresql.listen``
+        * ``postgresql.pg_hba`` and ``postgresql.pg_ident`` (see :func:`~RunningClusterConfigGenerator._set_conf_files`)
+
+        And redefine ``scope`` with the ``cluster_name`` GUC value if set.
+
+        :raises:
+            :class:`PatroniException`: if the provided user doesn't have superuser privileges.
+        """
+        self._set_su_params()
+
+        with self._get_connection_cursor() as cur:
+            self.pg_major = getattr(cur.connection, 'server_version', 0)
+
+            if not self._is_superuser(cur, self.config['postgresql']['authentication']['superuser']['username']):
+                raise PatroniException('The provided user does not have superuser privilege')
+
+            self._set_pg_params(cur)
+
+        self._set_conf_files()
+
+    def generate(self) -> None:
+        """Generate config using the info gathered from the specified running PG instance and update
+        `~RunningClusterConfigGenerator.config`."""
+        self._enrich_config_from_running_instance()
+        self.config['postgresql']['bin_dir'] = self._get_bin_dir_from_running_instance()
+        del self.config['bootstrap']['dcs']['standby_cluster']
+
+        self.merge_with_template()
```

Comment:
This two lines are exactly the same as here https://github.com/zalando/patroni/pull/2786/files#diff-cfee4a0506f1e512ce4c1fb59d7f7bac2f7924ad20e7d6a679044f4c5f4d6c88R187-R189

Also, self.generate() is called from both constructors independently. We probably can do something about it.